File: src/leetcode/P082_RemoveDuplicatesFromSortedListII.java
package leetcode;

/*
 * 	Given a sorted linked list, delete all nodes that have duplicate numbers,
 *  leaving only distinct numbers from the original list.

	For example,
	Given 1->2->3->3->4->4->5, return 1->2->5.
	Given 1->1->1->2->3, return 2->3.
 */

import tools.ListNode辅助.*;

/**
 * @author      zxwtry
 * @email       zxwtry@qq.com
 * @project     OJ
 * @package     leetcode
 * @file        P082_RemoveDuplicatesFromSortedListII.java
 * @date        2017年6月8日 下午10:24:44
 * @details     Solution: AC
 */
public class P082_RemoveDuplicatesFromSortedListII {
	static class Solution {
	    public ListNode deleteDuplicates(ListNode head) {
	    	if (head == null || head.next == null)
	    		return head;
	    	ListNode newHead = new ListNode(head.val == 0 ? -1 : 0);
	    	newHead.next = head;
	    	ListNode pre = newHead, cur = head.next;
	    	while (cur != null) {
	    		if (pre.next.val != cur.val) {
	    			if (pre.next.next == cur) {
	    				pre = pre.next;
	    				cur = cur.next;
	    			} else {
	    				pre.next = cur;
	    			}
	    		} else {
	    			cur = cur.next;
	    		}
	    	}
	    	if (pre.next.next != null)
	    		pre.next = null;
	        return newHead.next;
	    }
	}
}

File: python/P082_RemoveDuplicatesFromSortedListII.py
#coding=utf-8

'''
    url: leetcode.com/problems/remove-duplicates-from-sorted-list-ii
    @author:     zxwtry
    @email:      zxwtry@qq.com
    @date:       2017年4月20日
    @details:    Solution: 58ms 81.52%
'''

class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None
    def __repr__(self, *args, **kwargs):
        return str(self.val)

def construct(l):
    ln = len(l)
    ls = [None] * ln
    for i in range(ln-1, -1, -1):
        ls[i] = ListNode(l[i])
        if i != ln-1:
            ls[i].next = ls[i+1]
    return ls[0]
    
def print_list_node(l):
    print("==================")
    while l != None:
        print(l.val)
        l = l.next
    print("==================")
    
class Solution(object):
    def deleteDuplicates(self, h):
        """
        :type h: ListNode
        :rtype: ListNode
        """
        if h == None: return None
        t, p1, a, p2 = h, None, h, None
        while t != None:
            if p1 == None:
                p1 = t
            elif p1.val != t.val:
                if p1.next != t:
                    if p2 == None: a = t
                    else: p2.next = t
                else: p2 = p1
                p1 = t
            t = t.next
        if p1 != None and p1.next != None:
            if p2 != None: p2.next = None
            else: a = None
        return a

if __name__ == "__main__":
    #test case:
    #[1, 2, 2, 3]
    #[1, 2, 2, 3, 3, 3, 4, 4, 4, 5, 5, 6]
    #[2, 2, 3, 3, 4, 5, 5, 6, 6, 6, 7]
    #[1, 2, 2, 2, 2]
    #[2, 2, 3, 3, 4, 5, 5]
    #[2, 2, 3, 3, 4, 4, 5, 5]
    h = construct([2, 2, 3, 3, 4, 4, 5, 5])
    print_list_node(Solution().deleteDuplicates(h))